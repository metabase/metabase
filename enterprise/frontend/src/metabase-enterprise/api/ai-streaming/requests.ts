import { nanoid } from "@reduxjs/toolkit";

import api from "metabase/lib/api";

import { type AIStreamingConfig, processChatResponse } from "./process-stream";
import type { JSONValue } from "./types";

// keep track of inflight requests so that can be cancelled programmitcally from other
// places in the app w/o passing references to the abort controller directly
export const inflightAiStreamingRequests = new Map<
  string,
  { abortController: AbortController }
>();

export const getInflightRequestsForUrl = (url: string) => {
  return [...inflightAiStreamingRequests.entries()]
    .filter(([reqId]) => reqId.startsWith(`${url}-`))
    .map(([_reqId, reqInfo]) => reqInfo);
};

/**
 * Performs a streaming AI query by sending a POST request and processing the server-sent events response.
 * @returns A promise that resolves to the processed chat response stream
 */
export async function aiStreamingQuery(
  req: {
    url: string;
    signal?: AbortSignal | null | undefined;
    body: JSONValue;
  },
  config: AIStreamingConfig = {},
) {
  const reqId = `${req.url}-${nanoid()}`;

  try {
    const abortController = new AbortController();

    if (req.signal) {
      req.signal.addEventListener("abort", () => {
        abortController.abort(req.signal?.reason);
        inflightAiStreamingRequests.delete(reqId);
      });
    }
    inflightAiStreamingRequests.set(reqId, { abortController });

    const response = await fetch(req.url, {
      method: "POST",
      headers: {
        ...api.getClientHeaders(),
        Accept: "text/event-stream",
        "Content-Type": "application/json",
      },
      signal: abortController.signal,
      body: JSON.stringify(req.body),
    });

    if (!response.ok) {
      throw new Error(`Response status: ${response.status}`);
    }

    if (!response.body) {
      throw new Error("No response");
    }

    // without the await here the finally clause will run before
    // processChatResponse has resolved causing the chat response
    // to continue to be processed even if the provided abort
    // signal has been aborted
    return await processChatResponse(response.body, config);
  } finally {
    inflightAiStreamingRequests.delete(reqId);
  }
}
