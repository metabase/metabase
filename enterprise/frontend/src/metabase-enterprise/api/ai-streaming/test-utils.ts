import { defer } from "metabase/lib/promise";

async function delay(timeout: number) {
  return new Promise((resolve) => {
    setTimeout(() => resolve(undefined), timeout);
  });
}

export function createPauses<Count extends number>(count: Count) {
  const pauses = new Array(count).fill(null).map(() => defer());
  return pauses as ReturnType<typeof defer>[] & { length: Count };
}

export function createMockReadableStream(
  textChunks: string[] | AsyncGenerator<string, void, unknown>,
  options?: {
    disableAutoInsertNewLines: boolean;
  },
) {
  return new ReadableStream({
    async start(controller) {
      const textEncoder = new TextEncoder();
      try {
        for await (const textChunk of textChunks) {
          const text =
            textChunk + (options?.disableAutoInsertNewLines ? "" : "\n");
          controller.enqueue(textEncoder.encode(text));
        }
      } finally {
        controller.close();
      }
    },
  });
}

export function mockEndpoint<T extends Response>(
  url: string,
  endpointMock: () => Promise<T>,
) {
  const originalFetch = global.fetch;

  // fetch-mock is supposed to work with ReadableStreams, but when passed one
  // the getReader methods ends up as undefined
  return jest
    .spyOn(global, "fetch")
    .mockImplementation((fetchedUrl, ...args) => {
      const isRequestedUrl =
        typeof fetchedUrl === "string" && fetchedUrl.includes(url);

      if (isRequestedUrl) {
        return endpointMock();
      } else {
        return originalFetch(fetchedUrl, ...args);
      }
    });
}

export type MockStreamedEndpointParams =
  | {
      textChunks: string[] | undefined;
      stream?: undefined;
      initialDelay?: number;
    }
  | {
      textChunks?: undefined;
      stream: ReadableStream<any>;
      initialDelay?: number;
    };

export function mockStreamedEndpoint(
  url: string,
  { textChunks, stream, initialDelay = 0 }: MockStreamedEndpointParams,
) {
  return mockEndpoint(url, async () => {
    await delay(initialDelay);
    return {
      status: 202,
      ok: true,
      body:
        stream ||
        (textChunks && createMockReadableStream(textChunks)) ||
        undefined,
    } as any;
  });
}
