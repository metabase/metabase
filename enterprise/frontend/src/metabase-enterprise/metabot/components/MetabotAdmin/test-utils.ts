import type { MetabotId, SuggestedMetabotPrompt } from "metabase-types/api";

export const mockSuggestedPrompts: Record<MetabotId, SuggestedMetabotPrompt[]> =
  {
    "1": [
      {
        id: 1,
        metabot_id: 1,
        prompt: "What is the total revenue for this quarter?",
        model: "metric",
        model_id: 1,
        model_name: "Quarterly Revenue Calculator",
        created_at: "2025-05-15T10:30:00Z",
        updated_at: "2025-05-15T10:30:00Z",
      },
      {
        id: 2,
        metabot_id: 1,
        prompt:
          "Show me the customer acquisition trends over the last 6 months",
        model: "model",
        model_id: 2,
        model_name: "Customer Acquisition Trend Analyzer",
        created_at: "2025-05-15T11:15:00Z",
        updated_at: "2025-05-15T11:15:00Z",
      },
      {
        id: 3,
        metabot_id: 1,
        prompt: "What are our top performing products by sales volume?",
        model: "metric",
        model_id: 3,
        model_name: "Product Performance Ranking",
        created_at: "2025-05-15T14:22:00Z",
        updated_at: "2025-05-16T09:45:00Z",
      },
      {
        id: 4,
        metabot_id: 1,
        prompt: "How has our monthly recurring revenue changed this year?",
        model: "model",
        model_id: 1,
        model_name: "MRR Growth Tracker",
        created_at: "2025-05-16T08:30:00Z",
        updated_at: "2025-05-16T08:30:00Z",
      },
      {
        id: 5,
        metabot_id: 1,
        prompt: "What is our customer churn rate compared to last quarter?",
        model: "metric",
        model_id: 2,
        model_name: "Churn Rate Comparator",
        created_at: "2025-05-16T13:10:00Z",
        updated_at: "2025-05-16T13:10:00Z",
      },
      {
        id: 6,
        metabot_id: 1,
        prompt: "Show me the geographic distribution of our user base",
        model: "model",
        model_id: 4,
        model_name: "Geographic User Distribution Mapper",
        created_at: "2025-05-17T16:45:00Z",
        updated_at: "2025-05-18T10:20:00Z",
      },
      {
        id: 7,
        metabot_id: 1,
        prompt: "What is the average order value for new customers?",
        model: "metric",
        model_id: 4,
        model_name: "New Customer AOV Calculator",
        created_at: "2025-05-18T09:15:00Z",
        updated_at: "2025-05-18T09:15:00Z",
      },
      {
        id: 8,
        metabot_id: 1,
        prompt: "How do our conversion rates vary by traffic source?",
        model: "model",
        model_id: 3,
        model_name: "Traffic Source Conversion Analyzer",
        created_at: "2025-05-18T15:30:00Z",
        updated_at: "2025-05-18T15:30:00Z",
      },
    ],
    "2": [
      {
        id: 1,
        metabot_id: 1,
        prompt: "What is the total revenue for this quarter?",
        model: "metric",
        model_id: 1,
        model_name: "Quarterly Revenue Calculator",
        created_at: "2025-05-15T10:30:00Z",
        updated_at: "2025-05-15T10:30:00Z",
      },
    ],
  };
