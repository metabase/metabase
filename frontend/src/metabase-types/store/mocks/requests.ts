import type { RequestsState } from "../requests";

export const createMockRequestsState = (): RequestsState => {
  return {
    entities: {},
    plugins: {},
  };
};
