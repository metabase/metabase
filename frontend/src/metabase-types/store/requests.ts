export interface RequestState {
  loading: boolean;
  loaded: boolean;
  fetched: boolean;
  error: any | null;
  _isRequestState: true;
}
