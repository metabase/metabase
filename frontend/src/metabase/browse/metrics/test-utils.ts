import {
  createMockRecentCollectionItem,
  createMockSearchResult,
} from "metabase-types/api/mocks";

import type { MetricResult, RecentMetric } from "./types";

export const createMockMetricResult = (
  metric: Partial<MetricResult> = {},
): MetricResult =>
  createMockSearchResult({ ...metric, model: "metric" }) as MetricResult;

export const createMockRecentMetric = (
  metric: Partial<RecentMetric>,
): RecentMetric =>
  createMockRecentCollectionItem({
    ...metric,
    model: "metric",
  }) as RecentMetric;
