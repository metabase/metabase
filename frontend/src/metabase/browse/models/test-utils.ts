import type { RecentCollectionItem } from "metabase-types/api";
import {
  createMockRecentCollectionItem,
  createMockSearchResult,
} from "metabase-types/api/mocks";

import type { ModelResult, RecentModel } from "./types";

export const createMockModelResult = (
  model: Partial<ModelResult> = {},
): ModelResult =>
  createMockSearchResult({ ...model, model: "dataset" }) as ModelResult;

export const createMockRecentModel = (
  model: Partial<RecentCollectionItem>,
): RecentModel =>
  createMockRecentCollectionItem({ ...model, model: "dataset" }) as RecentModel;
