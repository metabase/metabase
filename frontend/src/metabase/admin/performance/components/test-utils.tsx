import { setupEnterprisePlugins } from "__support__/enterprise";
import { setupDatabasesEndpoints } from "__support__/server-mocks";
import { setupPerformanceEndpoints } from "__support__/server-mocks/performance";
import { mockSettings } from "__support__/settings";
import { createMockEntitiesState } from "__support__/store";
import { fireEvent, renderWithProviders, screen } from "__support__/ui";
import type { TokenFeatures } from "metabase-types/api";
import { DurationUnit } from "metabase-types/api";
import {
  createMockSettings,
  createMockTokenFeatures,
} from "metabase-types/api/mocks";
import {
  createMockCacheConfig,
  createMockCacheConfigWithDoNotCacheStrategy,
  createMockCacheConfigWithDurationStrategy,
  createMockCacheConfigWithTTLStrategy,
} from "metabase-types/api/mocks/performance";
import { createSampleDatabase } from "metabase-types/api/mocks/presets";
import { createMockState } from "metabase-types/store/mocks";

import { StrategyEditorForDatabases } from "./StrategyEditorForDatabases";

export interface SetupOpts {
  hasEnterprisePlugins?: boolean;
  tokenFeatures?: Partial<TokenFeatures>;
}

export const setup = ({
  hasEnterprisePlugins,
  tokenFeatures = {},
}: SetupOpts = {}) => {
  const storeInitialState = createMockState({
    entities: createMockEntitiesState({}),
    settings: mockSettings(
      createMockSettings({
        "token-features": createMockTokenFeatures(tokenFeatures),
      }),
    ),
  });

  if (hasEnterprisePlugins) {
    setupEnterprisePlugins();
  }

  const cacheConfigs = [
    createMockCacheConfigWithTTLStrategy({ model_id: 1 }),
    createMockCacheConfigWithDoNotCacheStrategy({ model_id: 2 }),
    createMockCacheConfigWithDurationStrategy({ model_id: 3 }),
    createMockCacheConfig({
      model: "root",
      model_id: 0,
      strategy: { type: "duration", duration: 1, unit: DurationUnit.Hours },
    }),
  ];
  setupPerformanceEndpoints(cacheConfigs);

  const databases = Array.from({ length: 4 }, (_, i) =>
    createSampleDatabase({ id: i + 1, name: `Database ${i + 1}`, tables: [] }),
  );
  setupDatabasesEndpoints(databases);

  return renderWithProviders(<StrategyEditorForDatabases />, {
    storeInitialState,
  });
};

export const getSaveButton = async () =>
  await screen.findByTestId("strategy-form-submit-button");

export const changeInput = async (
  label: RegExp,
  expectedPlaceholder: number,
  value: number,
) => {
  const input = (await screen.findByLabelText(label)) as HTMLInputElement;
  expect(input).toHaveAttribute("placeholder", expectedPlaceholder.toString());
  fireEvent.change(input, { target: { value } });
  expect(input).toHaveValue(value);
};
