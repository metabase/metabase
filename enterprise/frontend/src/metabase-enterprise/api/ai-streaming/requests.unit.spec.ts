import fetchMock from "fetch-mock";

import { aiStreamingQuery, getInflightRequestsForUrl } from "./requests";
import { mockStreamedEndpoint } from "./test-utils";

const ENDPOINT = "/some-streamed-endpoint";

describe("ai requests", () => {
  describe("aiStreamingQuery", () => {
    it("should return full result of a successful request", async () => {
      mockStreamedEndpoint(ENDPOINT, {
        textChunks: whoIsYourFavoriteResponse,
      });
      const result = await aiStreamingQuery({ url: ENDPOINT, body: {} });
      expect(result).toMatchSnapshot();
    });

    it("should call callbacks for relevant chunk types", async () => {
      mockStreamedEndpoint(ENDPOINT, {
        textChunks: [
          `0:"Testing"`,
          `2:{"type":"state","version":1,"value":{}}`,
          `9:{"toolCallId":"x","toolName":"x","args":""}`,
          `a:{"toolCallId":"x","result":""}`,
          `d:{"finishReason":"stop","usage":{"promptTokens":1,"completionTokens":1}}`,
        ],
      });

      const successCbs = {
        onTextPart: jest.fn(),
        onDataPart: jest.fn(),
        onToolCallPart: jest.fn(),
        onToolResultPart: jest.fn(),
        onError: jest.fn(),
      };

      await aiStreamingQuery({ url: ENDPOINT, body: {} }, successCbs);
      expect(successCbs.onTextPart).toHaveBeenCalled();
      expect(successCbs.onDataPart).toHaveBeenCalled();
      expect(successCbs.onToolCallPart).toHaveBeenCalled();
      expect(successCbs.onToolResultPart).toHaveBeenCalled();
      expect(successCbs.onError).not.toHaveBeenCalled();

      mockStreamedEndpoint(ENDPOINT, {
        textChunks: [
          `3:{}`, // error after finish to trigger all callbacks
        ],
      });

      const failureCbs = {
        onError: jest.fn(),
      };
      try {
        await aiStreamingQuery({ url: ENDPOINT, body: {} }, failureCbs);
      } catch (_) {}
      expect(failureCbs.onError).toHaveBeenCalled();
    });

    it("throw error if bad http status code", async () => {
      fetchMock.post(`path:${ENDPOINT}`, 500);
      await expect(
        aiStreamingQuery({ url: ENDPOINT, body: {} }),
      ).rejects.toThrow(/Response status: 500/);
    });

    it("throw error if no response", async () => {
      mockStreamedEndpoint(ENDPOINT, {
        textChunks: undefined,
      });
      await expect(
        aiStreamingQuery({ url: ENDPOINT, body: {} }),
      ).rejects.toThrow(/No response/);
    });

    it("should be able abort request via a passed in signal", async () => {
      const controller = new AbortController();

      fetchMock.post(`path:${ENDPOINT}`, { delay: 100, status: 200 });
      const promise = aiStreamingQuery({
        url: ENDPOINT,
        body: {},
        signal: controller.signal,
      });
      controller.abort();
      await expect(promise).rejects.toThrow(/The operation was aborted./);
    });

    describe("in-flight request tracking", () => {
      it("should register/unregister with inflight requests on a successful request", async () => {
        mockStreamedEndpoint(ENDPOINT, {
          textChunks: whoIsYourFavoriteResponse,
        });
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(0);
        const promise = aiStreamingQuery({ url: ENDPOINT, body: {} });
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(1);
        await promise;
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(0);
      });

      it("should properly unregister with inflight requests on abort", async () => {
        const controller = new AbortController();

        fetchMock.post(`path:${ENDPOINT}`, { delay: 100, status: 200 });
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(0);
        const promise = aiStreamingQuery({
          url: ENDPOINT,
          body: {},
          signal: controller.signal,
        });
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(1);
        controller.abort();
        expect(getInflightRequestsForUrl(ENDPOINT).length).toBe(0);
        try {
          await promise;
        } catch (err) {}
      });
    });
  });
});

const whoIsYourFavoriteResponse = [
  `0:"You, but don't tell anyone."`,
  `2:{"type":"state","version":1,"value":{"queries":{}}}`,
  `d:{"finishReason":"stop","usage":{"promptTokens":4916,"completionTokens":8}}`,
];
