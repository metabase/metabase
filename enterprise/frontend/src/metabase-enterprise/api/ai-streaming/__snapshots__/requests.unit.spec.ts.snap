// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ai requests aiStreamingQuery should return full result of a successful request 1`] = `
{
  "data": [
    {
      "type": "state",
      "value": {
        "queries": {},
      },
      "version": 1,
    },
  ],
  "history": [
    {
      "content": "You, but don't tell anyone.",
      "role": "assistant",
    },
  ],
  "parts": [
    {
      "code": "0",
      "name": "text",
      "value": "You, but don't tell anyone.",
    },
    {
      "code": "2",
      "name": "data",
      "value": {
        "type": "state",
        "value": {
          "queries": {},
        },
        "version": 1,
      },
    },
    {
      "code": "d",
      "name": "finish_message",
      "value": {
        "finishReason": "stop",
        "usage": {
          "completionTokens": 8,
          "promptTokens": 4916,
        },
      },
    },
  ],
  "text": "You, but don't tell anyone.",
  "toolCalls": [],
}
`;
